<a href="https://colab.research.google.com/github/spazznolo/optimize_pool_picks/blob/main/dsf_stc_exam.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [25]:
import pandas as pd
import numpy as np
import re, string

In [2]:
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

Outside of the target, which is binary (Disaster or not), all other relevant variables are characters. The main variable is the tweet itself, in the 'text' variable. Outside of that, there is a keyword variable (available for all but 0.8% of tweets) and a location variable (available for 66.7% of tweets).

In [24]:
train.isnull().mean(axis = 0)

id          0.000000
keyword     0.008013
location    0.332720
text        0.000000
target      0.000000
dtype: float64

33.3% of tweets have no location
00.8% of tweets have no keyword


In [4]:
train['text']

0       Our Deeds are the Reason of this #earthquake M...
1                  Forest fire near La Ronge Sask. Canada
2       All residents asked to 'shelter in place' are ...
3       13,000 people receive #wildfires evacuation or...
4       Just got sent this photo from Ruby #Alaska as ...
                              ...                        
7608    Two giant cranes holding a bridge collapse int...
7609    @aria_ahrary @TheTawniest The out of control w...
7610    M1.94 [01:04 UTC]?5km S of Volcano Hawaii. htt...
7611    Police investigating after an e-bike collided ...
7612    The Latest: More Homes Razed by Northern Calif...
Name: text, Length: 7613, dtype: object

In [5]:
# define a function to remove character references (e.g., &lt;, &amp;, &nbsp;)
def remove_references(text):
    text = re.sub(r'&[a-zA-Z]+;?', '', text)
    return text

# remove character references
train['text'] = train['text'].apply(remove_references)
test['text'] = test['text'].apply(remove_references)

In [6]:
# define a function to remove mentions
def remove_mentions(text):
    text = re.sub(r'@[a-zA-Z_]+;?', '', text)
    return text

# remove character references
train['text'] = train['text'].apply(remove_mentions)

In [7]:
train['text']

0       Our Deeds are the Reason of this #earthquake M...
1                  Forest fire near La Ronge Sask. Canada
2       All residents asked to 'shelter in place' are ...
3       13,000 people receive #wildfires evacuation or...
4       Just got sent this photo from Ruby #Alaska as ...
                              ...                        
7608    Two giant cranes holding a bridge collapse int...
7609      The out of control wild fires in California ...
7610    M1.94 [01:04 UTC]?5km S of Volcano Hawaii. htt...
7611    Police investigating after an e-bike collided ...
7612    The Latest: More Homes Razed by Northern Calif...
Name: text, Length: 7613, dtype: object

In [8]:
# define a function that removes numeric values and mixtures
def remove_num(text):
    text = re.sub(r'\w*\d+\w*', '', text)
    return text

# remove numeric values and mixtures
train['text'] = train['text'].apply(remove_num)

In [9]:
train['text']

0       Our Deeds are the Reason of this #earthquake M...
1                  Forest fire near La Ronge Sask. Canada
2       All residents asked to 'shelter in place' are ...
3       , people receive #wildfires evacuation orders ...
4       Just got sent this photo from Ruby #Alaska as ...
                              ...                        
7608    Two giant cranes holding a bridge collapse int...
7609      The out of control wild fires in California ...
7610         . [: UTC]? S of Volcano Hawaii. http://t.co/
7611    Police investigating after an e-bike collided ...
7612    The Latest: More Homes Razed by Northern Calif...
Name: text, Length: 7613, dtype: object

In [15]:
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer(stop_words='english', ngram_range=(1, 2))

In [16]:
X = vectorizer.fit_transform(train['text'])
df_bow_sklearn = pd.DataFrame(X.toarray(),columns=vectorizer.get_feature_names_out())
df_bow_sklearn

__  __ want  _one  _one gwatt  aa  aa ayyo  aa batteries  aa mgm  aaaa  \
0      0        0     0           0   0        0             0       0     0   
1      0        0     0           0   0        0             0       0     0   
2      0        0     0           0   0        0             0       0     0   
3      0        0     0           0   0        0             0       0     0   
4      0        0     0           0   0        0             0       0     0   
...   ..      ...   ...         ...  ..      ...           ...     ...   ...   
7608   0        0     0           0   0        0             0       0     0   
7609   0        0     0           0   0        0             0       0     0   
7610   0        0     0           0   0        0             0       0     0   
7611   0        0     0           0   0        0             0       0     0   
7612   0        0     0           0   0        0             0       0     0   

      aaaa ok  ...  ûóher  ûóher upper  ûókody  ûókody vine  ûónegligence  \
0           0  ...      0            0       0            0             0   
1           0  ...      0            0       0            0             0   
2           0  ...      0            0       0            0             0   
3           0  ...      0            0       0            0             0   
4           0  ...      0            0       0            0             0   
...       ...  ...    ...          ...     ...          ...           ...   
7608        0  ...      0            0       0            0             0   
7609        0  ...      0            0       0            0             0   
7610        0  ...      0            0       0            0             0   
7611        0  ...      0            0       0            0             0   
7612        0  ...      0            0       0            0             0   

      ûónegligence fireworks  ûótech  ûótech business  ûówe  ûówe work  
0                          0       0                0     0          0  
1                          0       0                0     0          0  
2                          0       0                0     0          0  
3                          0       0                0     0          0  
4                          0       0                0     0          0  
...                      ...     ...              ...   ...        ...  
7608                       0       0                0     0          0  
7609                       0       0                0     0          0  
7610                       0       0                0     0          0  
7611                       0       0                0     0          0  
7612                       0       0                0     0          0  

[7613 rows x 57908 columns]

In [40]:
X = X.toarray()
y = train['target']

In [44]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
#
# Create training and test split
#
X_train, X_test, y_train, y_test = train_test_split(X, y)
#
# Create an instance of LogisticRegression classifier
#
lr = LogisticRegression(random_state=1)
#
# Fit the model
#
lr.fit(X_train, y_train)
#
# Create the predictions
#
y_predict = lr.predict(X_test)
  
# Use metrics.accuracy_score to measure the score
print("LogisticRegression Accuracy %.3f" %metrics.accuracy_score(y_test, y_predict))

LogisticRegression Accuracy 0.811


In [19]:
# import TfidfVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

# instantiate the vectorizer
tfidf = TfidfVectorizer()

# fit and transform
X_tfidf = tfidf.fit_transform(train['text'])

# create a dataframe from the sparse matrix
X_tfidf_df = pd.DataFrame(X_tfidf.toarray())

# check the dataframe
X_tfidf_df.head()

0      1      2      3      4      5      6      7      8      9      ...  \
0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  ...   
1    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  ...   
2    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  ...   
3    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  ...   
4    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  ...   

   14774  14775  14776  14777  14778  14779  14780  14781  14782  14783  
0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
1    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
2    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
3    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
4    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  

[5 rows x 14784 columns]

In [21]:
X = X_tfidf_df
y = train['target']

In [22]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, f1_score, recall_score

clf = RandomForestClassifier(random_state=33, n_jobs=-1)
rf_model = clf.fit(X, y)
y_pred = rf_model.predict(X)

precision = precision_score(y, y_pred)
fscore = f1_score(y, y_pred)
train_accuracy = clf.score(X, y)
test_accuracy = accuracy_score(y, y_pred)

In [23]:
print(precision, fscore, train_accuracy, test_accuracy)

0.9928748451053284 0.9863055854746884 0.9883094706423223 0.9883094706423223
